py/dml/codegen.py

```diff
@@ -394,6 +395,462 @@ def memoization_common_prelude(name, site, outp, throws, make_ref):
                      make_ref('ran', TInt(8, True)),
                      mkCompound(site, unrun + has_run + running))]
 
+class TypeSequenceInfo:
+    """Bookkeeping surrounding a realtyped unique type sequence"""
+    def __init__(self, types, uniq):
+        self.types = types
+        self.uniq = uniq
+        self.after_on_hooks = {}
+        self.struct = (TStruct({f'comp{i}': typ
+                                 for (i, typ) in enumerate(types)},
+                                label=f'_typeseq_{self.uniq}')
+                       if types else None)
+        type_keys = ', '.join(f"'{typ.key()}'" for typ in types)
+        self.string_key = f'({type_keys})'
+
+    def get_after_on_hook(self, aoh_prim_key, param_to_msg_comp, no_params,
+                          createnew=False):
+        param_to_msg_comp_key = tuple(param_to_msg_comp.get(i)
+                                      for i in range(no_params))
+        try:
+            return self.after_on_hooks[(aoh_prim_key, param_to_msg_comp_key)]
+        except KeyError:
+            if not createnew:
+                raise
+            info = after_on_hook_info_constructors[type(aoh_prim_key)](
+                self, aoh_prim_key, param_to_msg_comp)
+            self.after_on_hooks[(aoh_prim_key, param_to_msg_comp_key)] = info
+            return info
+
+class TypeSequenceInfoDict:
+    """Mappings from a unique type sequence to its TypeSequenceInfo"""
+    def __init__(self):
+        self._dict = {}
+        self._list = []
+
+    def get(self, index, realtyped=False, createnew=False):
+        index = tuple(index if realtyped
+                      else (safe_realtype(typ) for typ in index))
+        classes = tuple(type(elem) for elem in index)
+        try:
+            infos = self._dict[classes]
+        except KeyError:
+            if createnew:
+                info = TypeSequenceInfo(index, len(self._list))
+                self._dict[classes] = [info]
+                self._list.append(info)
+                return info
+            else:
+                return None
+        for info in infos:
+            if (len(info.types) == len(index)
+                and all(typK.cmp(typI) == 0
+                        for (typK, typI) in zip(info.types, index))):
+                return info
+        if createnew:
+            info = TypeSequenceInfo(index, len(self._list))
+            self._list.append(info)
+            return info
+        else:
+            return None
+
+    def items(self):
+        return self._list
```

Comment:
`s/items/values` if you want to mimic the dict API. Also, why not `return (i for l in self._dict for i in l.values())`? That would allow `s/_list=[]/_num=0/`

Reply:
_I do not want to mimic the dict API_. I don't even want to make the attempt, because it's impossible get it completely right.

Reply:
Right, then `items` is still not good because it mimics (an unrelated part of) the dict API.

Reply:
Now irrelevant.